src/Components/Aspire.Elastic.Clients.Elasticsearch/ConfigurationSchema.json

```diff
@@ -0,0 +1,63 @@
+{
+  "type": "object",
+  "properties": {
+    "Aspire": {
+      "type": "object",
+      "properties": {
+        "Elastic": {
+          "type": "object",
+          "properties": {
+            "Clients": {
+              "type": "object",
+              "properties": {
+                "Elasticsearch": {
+                  "type": "object",
+                  "properties": {
+                    "Cloud": {
+                      "type": "object",
+                      "properties": {
+                        "ApiKey": {
```

Comment:
ApiKey authentication is not tied to Elastic Cloud, you can use API keys on self hosted Elasticsearch too:

https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html

Maybe create a separate `Authentication` configuration object that supports ApiKey and basic auth?

Reply:
In my opinion `Aspire.Elastic.Clients.Elasticsearch` is built for use in Aspire apps and `ElasticsearchResource` that is self-hosted uses password authentication right now, so if we want to support ApiKey in the client component, we should support ApiKey in hosting component too, right? 

I am not sure but I think if we want to support ApiKey in `ElasticsearchResource` we may face a problem in 
creating ConnectionString.

Reply:
The pattern we've been following in Aspire for connection information is to encode all the information needed to establish the connection into the `ConnectionString:<name>` configuration value, and then parsing the connection information from that. For example see:

https://github.com/dotnet/aspire/blob/6a4610c1d0d8213ecb3e5d0465256434dcb60852/src/Components/Aspire.Qdrant.Client/QdrantClientSettings.cs#L26-L47

https://github.com/dotnet/aspire/blob/6a4610c1d0d8213ecb3e5d0465256434dcb60852/src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs#L52-L72

Is this a pattern we can follow here? So these "Cloud" connection properties can be parsed from that string? Then in order to switch, you just need to update that string.

Reply:
> Is this a pattern we can follow here? So these "Cloud" connection properties can be parsed from that string? Then in order to switch, you just need to update that string.

Do you mean remove the `Cloud` section from `ElasticClientsElasticsearchSettings` or keep `Cloud` and retrieve ApiKey and CloudId from ConnectionString when the `Cloud` section is null?

by encoding information in ConnectionString we could support ApiKey in `ElasticsearchResource` too, it would be something like `Endpoint=Host:Port,ApiKey=someApiKey`, but do we need to support it in this PR?

Reply:
I think we could rework the Settings class a bit.

```csharp
public sealed class ElasticClientsElasticsearchSettings
{
    public Uri? Endpoint { get; set; }
    public string? Key { get; set; } // or ApiKey
    public string? CloudId { get; set; }
...
}
```

And then following the same pattern as shown in the above 2 existing libraries for breaking apart the `ConnectionStrings:<name>` string into these properties.

Then the question becomes what happens when both `Endpoint` and `CloudId` are specified? We can either throw an exception, or say one of them wins over the other.

> but do we need to support it in this PR?

I think we should get the initial design right, so we don't have to make potentially breaking changes in the future.

Reply:
> but do we need to support it in this PR?

To be clear - I don't think we need to support more scenarios than what is currently supported in this PR. I just want to make sure the public API is set up for future enhancements.

Reply:
> ApiKey authentication is not tied to Elastic Cloud, you can use API keys on self hosted Elasticsearch too

@Mpdreamz @stevejgordon - how is this possible with the .NET client? I only see these constructors:

* public ElasticsearchClientSettings(Uri? uri = null)
* public ElasticsearchClientSettings(string cloudId, AuthorizationHeader credentials)

How do you create an `ElasticsearchClient` using an API key on self hosted Elasticsearch?

Reply:
Since cloud requires it its part of the constructor but otherwise you can call `. Authentication()` on the settings itself for other scenarios.

Reply:
I Like the simplification of the settings so it projects 1 to 1 to a connection string.

I do think a single endpoint might not be sufficiently future proof for production deployments:

Scenario's

- Elastic CloudId. 
- Endpoints: []
- UseSniffing. 

If `Endpoint` is a SAAS offering a single one will be sufficient. 
However if it points to a self hosted Elasticsearch deployment we might want to support multiple endpoints.

Obviously in dev you would typically only launch a single Elasticsearch instance but in production you often have `>3` nodes to ensure availability. 

Elastic's transport layer can be seeded with multiple endpoints (e.g the master eligible nodes). Using sniffing the transport [can discover all the endpoints](https://github.com/elastic/elastic-transport-net/blob/eb5416e8df75e4c50215bf5258e13189afd9017d/src/Elastic.Transport/Products/Elasticsearch/ElasticsearchProductRegistration.cs#L103) that have data to be queried. This ensures the transport does not create hotspots and that it leaves nodes alone that should not be queried directly. See https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html for more about node roles. This way we play nicely with medium to large size deployments as well.

We could simplify this by always enabling sniffing OOTB if multiple endoints are specified.